sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessor.cs

```diff
@@ -681,16 +706,61 @@ public virtual async Task StopProcessingAsync(CancellationToken cancellationToke
-                                cancellationToken).
-                                ConfigureAwait(false);
+                            if (sessionId != null)
+                            {
+                                // Adding double-checked locking, which checks whether sessionIdsReceiverMap contains sessionId or not.
+                                // This is efficient because it avoids creating a receiver for the same session Ids for multiple threads.
+                                // Otherwise, we will get session lock lost error whenever multiple threads will try to create a receiver for
+                                // the same session id.
+                                if (!sessionIdsReceiverMap.ContainsKey(sessionId))
+                                {
+                                    await receiverCreationLockMap[sessionId].WaitAsync(cancellationToken).ConfigureAwait(false);
+                                    try
+                                    {
+                                        if (!sessionIdsReceiverMap.ContainsKey(sessionId))
+                                        {
+                                            receiver = await ServiceBusSessionReceiver.CreateSessionReceiverAsync(
+                                                EntityPath,
+                                                _connection,
+                                                 sessionId,
+                                                receiverOptions,
+                                                cancellationToken).ConfigureAwait(false);
+
+                                            sessionIdsReceiverMap.TryAdd(sessionId, receiver);
+
+                                            await OnSessionInitializingAsync(
+                                                (ServiceBusSessionReceiver)receiver,
+                                                cancellationToken).
+                                                ConfigureAwait(false);
+                                        }
+                                        else
+                                        {
+                                            receiver = sessionIdsReceiverMap[sessionId];
+                                        }
+                                    }
+                                    finally
+                                    {
+                                        receiverCreationLockMap[sessionId].Release();
+                                    }
+                                }
+                                else
+                                {
+                                    receiver = sessionIdsReceiverMap[sessionId];
+                                }
+                            }
+                            else
+                            {
+                                receiver = await ServiceBusSessionReceiver.CreateSessionReceiverAsync(
```

Comment:
Consider creating a separate method, since we always need to do both of these things at the same time.